### ISSUE TITLE
memory leak with sqlx.Select

### ISSUE THREAD

**Issue by volkmarbuehringer** [rana/ora#160]
I have clearly visible memory leak while fetching data from oracle
into an slice with sqlx.Select

the same code works with Postgres without problems

		if err = db.DB.Select(m[0], sql, id); err != nil {
			checkErr(err)
		}

**Comment by tgulacsi** [comment]
Please help me!

What does sqlx.Select do exactly? What is the table, what kind of columns does it have, what is the rate of the memory leak?
And what version of rana/ora  are you using?

**Comment by volkmarbuehringer** [comment]
I compared your driver with the postgres driver, which is written in go.
The pg-driver uses only a fraction of memory and performs much faster.
But I think it is because of the oci-libraries you are using with c-go

The memory consumption with oracle stabilizes at 90MB and the server fetches 2,5 MB /sec from the DB (serving of json )
The pg-driver fetches 6.5MB/sec and uses 18MB of memory.

I tried another code-sample in which the closing of the result-set is done by me, but this changes nothing at the memory consumption:

		rows, err1 := db.DB.Queryx(`select * from lt_ads where ads_id > :1 and rownum < 100`, id)
		checkErr(err1)
		defer rows.Close()
		results := make(dbRows, 0, 100)
		for rows.Next() {
			t := make(map[string]interface{})
			checkErr(rows.MapScan(t))
			results = append(results, t)
		}
		jsonString, err := json.Marshal(results)
		checkErr(err)

		ctx.Writef(string(jsonString))

**Comment by tgulacsi** [comment]
The default prefetchRowCount is 128, the max prefetch memory is 128MiB, (C
side) plus the driver fetches columns in 128 batches (Go side) by default.
This should explain the bigger memory usage, at least part of.

The speed: cgo (calling C from go) is not fast, and I think Go's network
code is faster than Oracle OCI's.
Though maybe we can work on it you help profile & benchmark the bottleneck.

volkmarbuehringer <notifications@github.com> ezt írta (időpont: 2017. febr.
23., Cs, 16:22):

I compared your driver with the postgres driver, which is written in go.
The pg-driver uses only a fraction of memory and performs much faster.
But I think it is because of the oci-libraries you are using with c-go

The memory consumption with oracle stabilizes at 90MB and the server
fetches 2,5 MB /sec from the DB (serving of json )
The pg-driver fetches 6.5MB/sec and uses 18MB of memory.

I tried another code-sample in which the closing of the result-set is done
by me, but this changes nothing at the memory consumption:

	rows, err1 := db.DB.Queryx(`select * from lt_ads where ads_id > :1
and rownum < 100`, id)
	checkErr(err1)
	defer rows.Close()
	results := make(dbRows, 0, 100)
	for rows.Next() {
		t := make(map[string]interface{})
		checkErr(rows.MapScan(t))
		results = append(results, t)
	}
	jsonString, err := json.Marshal(results)
	checkErr(err)

	ctx.Writef(string(jsonString))

—
You are receiving this because you commented.


Reply to this email directly, view it on GitHub
<https://github.com/rana/ora/issues/160#issuecomment-282021725>, or mute
the thread
